draft-irtf-cfrg-opaque.md

```diff
@@ -1665,6 +1663,64 @@ learn the client's password. This means that additional checks should be conside
+| InvalidClientEphemeralPublicKeyError | Server | Ke1                       | Invalid ephemeral client public key.                 |
+| InvalidServerEphemeralPublicKeyError | Client | Ke2                       | Invalid ephemeral server public key.                 |
+
+### Client errors {#client-errors}
+
+The client can produce errors due to incorrect values in the messages it received. The following
+table enumerates some of these errors, where they occur, and the reason for the error.
+
+| Name                        | Stage                                                  | Reason                                                                   |
+|:----------------------------|:-------------------------------------------------------|:-------------------------------------------------------------------------|
+| InvalidInputError           | CreateRegistrationRequest,<br/>CreateCredentialRequest | OPRF input deterministically maps to the group identity element.         |
+| InvalidMaskedLengthError    | ClientFinish                                           | The length of the masked response is not = point length + envelope size. |
+| InvalidServerPublicKeyError | ClientFinish                                           | Invalid server public key in unmasked response.                          |
+| InvalidEnvelopeMACError     | ClientFinish                                           | Invalid envelope authentication tag.                                     |
+| InvalidServerAkeMACError    | ClientFinish                                           | Invalid server AKE MAC.                                                  |
+
+
+### Server errors {#server-errors}
+
+The following table enumerates server errors that can occur, where they occur, and the reason for the error.
+In this case, servers can only generate a single protocol error.
+
+| Name                     | Stage        | Reason                  |
+|:-------------------------|:-------------|:------------------------|
+| InvalidClientAkeMACError | ServerFinish | Invalid client AKE MAC. |
+
+### Application errors
+
+Implementations MUST verify certain input arguments and return an appropriate error. Note that
+these are not protocol errors, and are indicative. The following table enumerates server errors that can occur, where they occur, and the reason for the error.
+
+| Name                        | Role   | Stage      | Reason                                         |
+|:----------------------------|:-------|:-----------|:-----------------------------------------------|
+| InvalidServerSecretKeyError | Server | LoginInit  | Invalid server secret key.                     |
+| InvalidServerPublicKeyError | Server | LoginInit  | Invalid server public key.                     |
+| InvalidOPRFSeedLengthError  | Server | LoginInit  | Invalid OPRF seed length.                      |
```

Comment:
Why would this be thrown from the protocol, rather than enforced as a pre-condition to LoginInit? (Also, what is the LoginInit stage? It's not listed in the doc.)

Reply:
Yes, I wanted to discuss this in the PR-draft stage. As stated, implementations must observe these conditions, and can raise these errors, since these conditions are necessary for the protocol to execute correctly

Reply:
I suppose LoginInit comes from my implementation, my bad! Amending right now